lisa/sut_orchestrator/hyperv/source.py

```diff
@@ -0,0 +1,190 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+import uuid
+from pathlib import Path, PureWindowsPath
+from typing import List, Optional, Type
+
+from lisa import schema
+from lisa.node import RemoteNode
+from lisa.tools.cp import Cp
+from lisa.tools.ls import Ls
+from lisa.tools.mkdir import Mkdir
+from lisa.tools.powershell import PowerShell
+from lisa.tools.rm import Rm
+from lisa.tools.unzip import Unzip
+from lisa.util import InitializableMixin, subclasses
+from lisa.util.logger import get_logger
+
+from .schema import LocalSourceSchema, SourceFileSchema, SourceSchema
+
+
+class Source(subclasses.BaseClassWithRunbookMixin, InitializableMixin):
+    def __init__(self, runbook: SourceSchema) -> None:
+        super().__init__(runbook=runbook)
+        self._log = get_logger("source", self.__class__.__name__)
+
+    @classmethod
+    def type_schema(cls) -> Type[schema.TypedSchema]:
+        return SourceSchema
+
+    def download(self, server: RemoteNode) -> List[PureWindowsPath]:
+        raise NotImplementedError()
+
+
+class LocalSource(Source):
+    def __init__(self, runbook: LocalSourceSchema) -> None:
+        super().__init__(runbook)
+        self.local_runbook: LocalSourceSchema = self.runbook
+        self._log = get_logger("local", self.__class__.__name__)
+        self.downloaded_files: Optional[List[PureWindowsPath]] = None
+
+    @classmethod
+    def type_name(cls) -> str:
+        return "local"
+
+    @classmethod
+    def type_schema(cls) -> Type[schema.TypedSchema]:
+        return LocalSourceSchema
+
+    def download(self, server: RemoteNode) -> List[PureWindowsPath]:
+        if not self.downloaded_files:
+            self.downloaded_files = []
+
+        for file in self.local_runbook.files:
+            self.downloaded_files += self._download_file(file, server)
+
+        return self.downloaded_files
+
+    def _download_file(
+        self, file: SourceFileSchema, server: RemoteNode
+    ) -> List[PureWindowsPath]:
+        local_path = Path(file.source)
+
+        local_files = []
+        if local_path.is_dir():
+            local_files = [f for f in local_path.iterdir() if f.is_file()]
+        else:
+            local_files = [local_path]
+
+        downloaded_paths: List[PureWindowsPath] = []
+
+        for f in local_files:
+            downloaded_paths += self._download_single_file(
+                f, file.destination, file.unzip, server
+            )
+
+        return downloaded_paths
+
+    def _download_single_file(
+        self,
+        local_path: Path,
+        destination: Optional[str],
+        unzip: bool,
+        server: RemoteNode,
+    ) -> List[PureWindowsPath]:
+        if unzip:
+            return self._download_zipped_file(local_path, destination, server)
```

Comment:
The logic can be 1) copy, 2) unzip. So the copy logic is sharable, regardless unzip or not.

Reply:
After assuming destination is a dir, the code has simplified. Please review again.